platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/VirtualRootNode.java

```diff
@@ -477,42 +479,80 @@ public void fullLeafRehash() {
                 .build()
                 .start();
 
-        for (long i = firstLeafPath; i <= lastLeafPath; i++) {
-
-            try {
-                VirtualLeafRecord<K, V> leafRecord = dataSource.loadLeafRecord(i);
-                if (leafRecord != null) {
-                    try {
-                        final boolean success =
-                                rehashIterator.supply(leafRecord, MAX_FULL_REHASHING_BUFFER_TIMEOUT, SECONDS);
-                        if (!success) {
-                            throw new MerkleSynchronizationException(
-                                    "Timed out waiting to supply a new leaf to the hashing iterator buffer");
+        // This background thread will be responsible for feeding the iterator with data.
+        new ThreadConfiguration(getStaticThreadManager())
+                .setComponent("virtualmap")
+                .setThreadName("leafFeeder")
+                .setRunnable(() -> {
+                    long onePercent = (lastLeafPath - firstLeafPath) / 100 + 1;
+                    for (long i = firstLeafPath; i <= lastLeafPath; i++) {
+                        try {
+                            VirtualLeafRecord<K, V> leafRecord = dataSource.loadLeafRecord(i);
+                            if (leafRecord != null) {
```

Comment:
Can it actually be null? To me, all records in [firstLeafPath, lastLeafPath] must be resolvable at the data source level. An assertion (or even better, exception) would be beneficial here

Reply:
sounds good, I'll replace it with an assert.